### PR TITLE
fix(aws_lambda): fix circular dependency when user handler patches modules

### PR DESCRIPTION
**Pull request by duncanista**
Initially discovered in [datadog-lambda-python#313](https://github.com/DataDog/datadog-lambda-python/issues/313).

When patching, code can contain `ddtrace.patch_all()` or `ddtrace.patch(aws_lambda=True)`. Creating a circular dependency, triggering another patch. To avoid this, we skip on `AttributeError` which is the error thrown when patching a second time.

This was only affecting users patching again on their AWS Lambda functions instrumented by Datadog.

## Testing strategy

Tested on AWS Lambda functions. It was not possible to reproduce it in unit testing.

## Risk

There might be a better way to figure out when a circular dependency is happening.

## Checklist

- [x] Change(s) are motivated and described in the PR description.
- [x] Testing strategy is described if automated tests are not included in the PR.
- [x] Risk is outlined (performance impact, potential for breakage, maintainability, etc).
- [x] Change is maintainable (easy to change, telemetry, documentation).
- [x] [Library release note guidelines](https://ddtrace.readthedocs.io/en/stable/contributing.html#Release-Note-Guidelines) are followed.
- [x] Documentation is included (in-code, generated user docs, [public corp docs](https://github.com/DataDog/documentation/)).
- [x] Author is aware of the performance implications of this PR as reported in the benchmarks PR comment.

## Reviewer Checklist

- [x] Title is accurate.
- [x] No unnecessary changes are introduced.
- [x] Description motivates each change.
- [x] Avoids breaking [API](https://ddtrace.readthedocs.io/en/stable/versioning.html#interfaces) changes unless absolutely necessary.
- [x] Testing strategy adequately addresses listed risk(s).
- [x] Change is maintainable (easy to change, telemetry, documentation).
- [x] Release note makes sense to a user of the library.
- [x] Reviewer is aware of, and discussed the performance implications of this PR as reported in the benchmarks PR comment.

### CODE REVIEW

File: ddtrace/contrib/aws_lambda/patch.py [hunk 1]
```diff
@@ -165,23 +165,37 @@ def patch():
...
-    handler, handler_module, wrapper = _get_handler_and_module()
+    try:
+        handler, handler_module, wrapper = _get_handler_and_module()
 
-    if getattr(handler_module, "_datadog_patch", False):
-        return
-    setattr(handler_module, "_datadog_patch", True)
+        if getattr(handler_module, "_datadog_patch", False):
+            return
+        setattr(handler_module, "_datadog_patch", True)
```
Comment (joeyzhao2018) [inline ddtrace/contrib/aws_lambda/patch.py:173]: Should we set the attr after wrap()?

Comment (duncanista) [reply ddtrace/contrib/aws_lambda/patch.py:173]: I think the order here doesn't matter, but what do you think?

Comment (joeyzhao2018) [reply ddtrace/contrib/aws_lambda/patch.py:173]: if the `wrap()` fail, it will stil set attr `_datadog_path` to True, right?

Comment (duncanista) [reply ddtrace/contrib/aws_lambda/patch.py:173]: You're right, i'll switch the spots of the unpatch too.

File: ddtrace/contrib/aws_lambda/patch.py [hunk 2]
```diff
@@ -165,23 +165,37 @@ def patch():
...
         return
 
-    handler, handler_module, wrapper = _get_handler_and_module()
+    try:
+        handler, handler_module, wrapper = _get_handler_and_module()
 
-    if getattr(handler_module, "_datadog_patch", False):
-        return
-    setattr(handler_module, "_datadog_patch", True)
+        wrap(handler, wrapper)
```
Comment (joeyzhao2018) [inline ddtrace/contrib/aws_lambda/patch.py:171]: Actually, I was thinking this:
```
if getattr(handler_module, "_datadog_patch", False):
    return
wrap(handler, wrapper)        
setattr(handler_module, "_datadog_patch", True)
```

File: releasenotes/notes/fix-aws-lambda-circular-dep-when-patching-in-handler-320de20015a8d5cd.yaml [hunk 3]
```diff
@@ -0,0 +1,4 @@
+---
+fixes:
+  - |
+    aws_lambda: Fix AttributeError raised when `ddtrace.patch_all()`, or `ddtrace.patch(aws_lambda=True)`, is set on user handler.
```
Comment (ZStriker19) [inline releasenotes/notes/fix-aws-lambda-circular-dep-when-patching-in-handler-320de20015a8d5cd.yaml:4]: can we add a newline character here?
```suggestion
    aws_lambda: Fix AttributeError raised when `ddtrace.patch_all()`, or `ddtrace.patch(aws_lambda=True)`, is set on user handler.
    
```

**Comment by P403n1x87** [comment]
Is it not possible to break the circular import in the first place?

**Comment by duncanista** [comment]
> Is it not possible to break the circular import in the first place?

Yes, once we stop being a contribution, and patching users code. 

On the current code, I tried marking the code as already patched before doing it, but it didn't show a difference, will try doing it again to see if it changes again, but currently, I couldn't find any other workaround.

**Review by joeyzhao2018 — DISMISSED** [review]
LGTM.

**Review by emmettbutler — DISMISSED** [review]

**Review by emmettbutler — APPROVED** [review]

**Pull request merged**